File: easy/2200. Find All K-Distant Indices in an Array/2200. Find All K-Distant Indices in an Array.py
from typing import List


class Solution:
    def findKDistantIndices(self, nums: List[int], key: int, k: int) -> List[int]:
        return self.solution2(nums, key, k)

    def solution1(self, nums: List[int], key: int, k: int) -> List[int]:
        res = set()
        for i, num in enumerate(nums):
            if num == key:
                for j in range(max(0, i - k), min(len(nums), i + k + 1), 1):
                    res.add(j)
        return list(res)

    def solution2(self, nums: List[int], key: int, k: int) -> List[int]:
        n = len(nums)
        res = set()

        for i, num in enumerate(nums):
            if num == key:
                start = max(0, i - k)
                end = min(n - 1, i + k)
                for j in range(start, end + 1):
                    res.add(j)

        return list(res)


solution = Solution()
print(solution.findKDistantIndices(nums=[3, 4, 9, 1, 3, 9, 5], key=9, k=1))
print(solution.findKDistantIndices(nums=[2, 2, 2, 2, 2], key=2, k=2))

File: easy/2200. Find All K-Distant Indices in an Array/2200. Find All K-Distant Indices in an Array.ts
function findKDistantIndices(nums: number[], key: number, k: number): number[] {
    const res: number[] = []
    const n = nums.length
    let r = 0

    for (let j = 0; j < n; j++) {
        if (nums[j] === key) {
            const l = Math.max(r, j - k)
            r = Math.min(n - 1, j + k) + 1

            for (let i = l; i < r; i++)
                res.push(i)
        }
    }

    return res
};